### PR TITLE
[jit][script] Make empty list literals construct empty Tensor[]

### PR DESCRIPTION
**Pull request by suo**
This will make the common case more natural (no need to do `_construct_empty_tensor_list()`)

### CODE REVIEW

**Comment by michaelsuo** [comment]
I'm concerned that the error msg is confusing:
```
variable 'x' previously has type Tensor[] but is now being assigned to a value of type int[]:
def reassign_from_empty_literal():
    x = []
    if True:
        x = [1, 2, 3]
        ~ <--- HERE
    return
```
Should I add special casing for reporting this error? (e.g. if we're trying to assign to a tensor list and get a type error, mention that empty lists are constructed as tensor lists).

File: torch/csrc/jit/script/compiler.cpp [hunk 1]
```diff
@@ -1375,9 +1388,9 @@ struct to_ir {
         auto ll = ListLiteral(tree);
         auto values = getValues(ll.inputs(), /*maybe_unpack=*/true, identity);
         if (values.size() == 0) {
-          throw ErrorReport(tree) << "Empty list literals not allowed. "
-                                  << "Use _construct_empty_foo_list() instead. "
-                                  << "`foo` can be `int`, `float` or `tensor`";
+          // If this is an empty list literal `[]`, construct an empty Tensor[]
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: nit: less code duplication with:
```
auto elem_type = values.size() == 0 ? DynamicType::get() :  values.at(0)->type();
```

**Review by zdevito — APPROVED** [review]
Looks good. Custom error is useful until we get this working better.

File: torch/csrc/jit/script/compiler.cpp [hunk 2]
```diff
@@ -254,9 +254,22 @@ struct Environment {
...
+        errMsg << "variable '" << name << "' previously has type "
+               << simple_parent->type()->str()
+               << " but is now being assigned to a value of type "
+               << as_simple_value->type()->str();
+        // Special-cased error msg if we're trying to assign to a tensor list.
+        if (simple_parent->type()->kind() == TypeKind::ListType &&
+            as_simple_value->type()->kind() == TypeKind::ListType) {
+          errMsg << "\n. (Note: empty lists are constructed as Tensor[]; "
+                 << "if you want an empty list of a different type, "
+                 << "use `_construct_empty_foo_list`, "
```
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: Can we please at least call it `torch.jit.empty_int_list`? 😬

**Pull request closed without merging**